examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py

```diff
@@ -0,0 +1,624 @@
+
+    # Define gradient update step fn
+    def train_step(state, batch, dropout_rng):
+        dropout_rng, new_dropout_rng = jax.random.split(dropout_rng)
+
+        def loss_fn(params):
+            labels = batch.pop("labels")
+
+            logits = state.apply_fn(**batch, params=params, dropout_rng=dropout_rng, train=True)[0]
+
+            # compute loss, ignore padded input tokens
+            label_mask = jnp.where(labels > 0, 1.0, 0.0)
+            loss = optax.softmax_cross_entropy(logits, onehot(labels, logits.shape[-1])) * label_mask
+
+            # take average
+            loss = loss.sum() / label_mask.sum()
+
+            return loss
+
+        grad_fn = jax.value_and_grad(loss_fn)
+        loss, grad = grad_fn(state.params)
+        grad = jax.lax.pmean(grad, "batch")
+        new_state = state.apply_gradients(grads=grad)
+
+        metrics = jax.lax.pmean(
+            {"loss": loss, "learning_rate": linear_decay_lr_schedule_fn(state.step)}, axis_name="batch"
+        )
+
+        return new_state, metrics, new_dropout_rng
+
+    # Create parallel version of the train step
+    p_train_step = jax.pmap(train_step, "batch", donate_argnums=(0,))
+
+    # Define eval fn
+    def eval_step(params, batch):
+        labels = batch.pop("labels")
+
+        logits = model(**batch, params=params, train=False)[0]
+
+        # compute loss, ignore padded input tokens
+        label_mask = jnp.where(labels > 0, 1.0, 0.0)
+        loss = optax.softmax_cross_entropy(logits, onehot(labels, logits.shape[-1])) * label_mask
+
+        # compute accuracy
+        accuracy = jnp.equal(jnp.argmax(logits, axis=-1), labels) * label_mask
+
+        # summarize metrics
+        metrics = {"loss": loss.sum(), "accuracy": accuracy.sum(), "normalizer": label_mask.sum()}
+        metrics = jax.lax.psum(metrics, axis_name="batch")
+
+        return metrics
+
+    p_eval_step = jax.pmap(eval_step, "batch", donate_argnums=(0,))
+
+    # Replicate the train state on each device
+    state = jax_utils.replicate(state)
+
+    train_time = 0
+    train_start = time.time()
+    train_metrics = []
+    eval_metrics = []
+
+    training_iter = iter(tokenized_datasets)
+
+    max_seq_length = min(data_args.max_seq_length, tokenizer.model_max_length)
+    training_iter, eval_samples = advance_iter_and_group_samples(
+        training_iter, data_args.num_eval_samples, max_seq_length
+    )
+
+    steps = tqdm(range(num_train_steps), desc=f"Training...", position=0)
+    for step in range(num_train_steps):
+        # ======================== Training ================================
+        try:
+            training_iter, samples = advance_iter_and_group_samples(training_iter, train_batch_size, max_seq_length)
+        except StopIteration:
+            # Once the end of the dataset stream is reached, the training iterator
+            # is reinitialized and reshuffled and a new eval dataset is randomely chosen.
```

Comment:
is there any way to fix the eval dataset, changing the eval dataset would make it difficult to track the metrics no?

Reply:
good point! IMO, this script should be used only for huge datasets where most of the times running for a single epoch is enough, so that this doesn't really matter. 

@lhoestq - what do you think?

Reply:
Once this PR is merged https://github.com/huggingface/datasets/pull/2582 you could do something like
```python
dataset = load_dataset(..., streaming=True).shuffle(...)

eval_size = 10000  # feel free to change the size
eval_dataset = dataset.take(eval_size)
train_dataset = dataset.skip(eval_size).shuffle(...)
```